contracts/clients/src/BlsProvider.ts

```diff
@@ -0,0 +1,155 @@
+import { ethers, BigNumber } from "ethers";
+import { parseEther, Deferrable } from "ethers/lib/utils";
+
+import { ActionData, Bundle } from "./signer/types";
+import Aggregator, { BundleReceipt } from "./Aggregator";
+import BlsSigner, { _constructorGuard } from "./BlsSigner";
+import poll from "./helpers/poll";
+
+export default class BlsProvider extends ethers.providers.JsonRpcProvider {
+  readonly aggregator: Aggregator;
+  readonly verificationGatewayAddress: string;
+  signer!: BlsSigner;
+
+  constructor(
+    aggregatorUrl: string,
+    verificationGatewayAddress: string,
+    url?: string,
+    network?: ethers.providers.Networkish,
+  ) {
+    super(url, network);
+    this.aggregator = new Aggregator(aggregatorUrl);
+    this.verificationGatewayAddress = verificationGatewayAddress;
+  }
+
+  override async estimateGas(
+    transaction: Deferrable<ethers.providers.TransactionRequest>,
+  ): Promise<BigNumber> {
+    if (!transaction.to) {
+      throw new TypeError("Transaction.to should be defined.");
+    }
+
+    try {
+      const action: ActionData = {
+        ethValue: transaction.value?.toString() ?? "0",
+        contractAddress: transaction.to.toString(),
+        encodedFunction: transaction.data?.toString() ?? "0x",
+      };
+
+      const signer = this.getSigner();
+      const bundle = await signer.signBlsTransaction(action);
+
+      const gasEstimate = await this.aggregator.estimateFee(bundle);
+
+      return parseEther(gasEstimate.feeRequired);
+    } catch (error) {
+      throw new Error(`estimateGas() - an unexpected error occured: ${error}`);
+    }
+  }
+
+  override async sendTransaction(
+    signedTransaction: string | Promise<string>,
+  ): Promise<ethers.providers.TransactionResponse> {
+    throw new Error(
+      "sendTransaction() is not implemented. Call 'sendBlsTransaction()' instead.",
+    );
+  }
```

Comment:
Much like we do in the extension, we should parse/decode the signed txn and convert it to a bundle internally.

Reply:
As mentioned in another comment, have removed the extra methods I added for BLS logic, and converted/moved the logic into the correct ethers methods